### PR TITLE
Circuitpython nickzoic 1042 nrf nvm bytearray 2

### PR DESCRIPTION
**Pull request by nickzoic**
OK, so this is a new pull request at last for #1042, replacing #1610 which I've closed.

This version moves the code for flash access into `ports/nrf/peripherals/nrf/nvm.c` which is used by both `ports/nrf/supervisor/internal_flash.c` and by `ports/nrf/common-hal/nvm/ByteArray.c`.

The NVM store uses page(s) at the end of the FATFS page allocation, depending on the size of CIRCUITPY_INTERNAL_NVM_SIZE.

### CODE REVIEW

File: ports/nrf/supervisor/internal_flash.c [hunk 1]
```diff
@@ -68,81 +67,14 @@ uint32_t supervisor_flash_get_block_size(void) {
...
-    }
-    return sd_flash_operation_status;
-}
-#endif
-
-void supervisor_flash_flush(void) {
-    if (_flash_page_addr == NO_CACHE) return;
-
-    // Skip if data is the same
-    if (memcmp(_flash_cache, (void *)_flash_page_addr, FL_PAGE_SZ) != 0) {
```
Comment (dhalbert) [inline ports/nrf/supervisor/internal_flash.c]: This old code checked for writing identical contents to the flash page, and skipped if that were true. I don't see the same check in your refactored code.

File: ports/nrf/supervisor/internal_flash.c [hunk 2]
```diff
@@ -68,81 +67,14 @@ uint32_t supervisor_flash_get_block_size(void) {
...
-
-STATIC sd_flash_operation_status_t sd_flash_operation_wait_until_done(void) {
-    while (sd_flash_operation_status == SD_FLASH_OPERATION_IN_PROGRESS) {
-        sd_app_evt_wait();
-    }
-    return sd_flash_operation_status;
-}
-#endif
-
-void supervisor_flash_flush(void) {
```
Comment (dhalbert) [inline ports/nrf/supervisor/internal_flash.c]: `supervisor_flash_flush()` is called from `supervisor/shared/filesystem.c`, but I don't see it in your refactored code.

**Comment by dhalbert** [comment]
Also, could you make similar changes to the pca10056 build so it can use nvm?

File: ports/nrf/supervisor/internal_flash.c (hunk 1 above)
Comment (nickzoic) [reply ports/nrf/supervisor/internal_flash.c]: Sorry, managed to break that while rebasing it I think ... have just pushed changes which puts it back together, and now I'm making clean and rerunning my tests!

File: ports/nrf/supervisor/internal_flash.c (hunk 2 above)
Comment (nickzoic) [reply ports/nrf/supervisor/internal_flash.c]: Yep, see above :-(

**Comment by nickzoic** [comment]
> See comments. Also BOARD=pca10059 doesn't build.

Yep, I've added in a catch this so that if CIRCUITPY_INTERNAL_NVM_SIZE isn't defined it defaults to zero.

**Comment by nickzoic** [comment]
OK so I also modified your `adafruit_ble.uart.UARTServer` code to write to flash while it was talking and that passes tests too so hopefully ...

**Comment by dhalbert** [comment]
@nickzoic - hi, I started reviewing this yesterday, but was getting errors when trying to subscript nvm on a Feather 52840. And then I ran into #1773, which interfered with debugging. Will go back to testing this todya.

**Comment by tannewt** [comment]
Closing in favor of #1779. Thanks for getting us going!

**Pull request closed without merging**